### ISSUE TITLE
How to resume training from checkpoint in HF format?

### ISSUE THREAD

**Issue by zzz1YAO** [volcengine/verl#2767]
I'm a beginner and have a beginner's question about using GRPO.
When I set `trainer.save_freq=50`, I get checkpoints in both FSDP format and HuggingFace format. I then convert them to HF model files.
Now I want to use this checkpoint to continue training. I know I can set:
```
trainer.resume_from_path=xxx
trainer.resume_mode=resume_from_path
```
But I have two questions:

Which directory path should I use for resume_from_path?

The HuggingFace model path, or
The FSDP model file path?

What's the difference between using the checkpoint model file in:

`trainer.resume_from_path` vs
`actor_rollout_ref.model.path?
`

Thanks for anyone's answer !!!

**Comment by vermouth1992** [comment]
There is no way to "resume" from hf ckpt as it is meant to start from scratch. You can only resume from what ckpt manager saves.

**Comment by zzz1YAO** [comment]
> There is no way to "resume" from hf ckpt as it is meant to start from scratch. You can only resume from what ckpt manager saves.

Thank you for your reply!

I'm saving with FSDP, and because the weight files for training a 7B model are quite large, I've only saved the model files themselves, without the optimizer or extra states. Will this cause the resume to fail?

Also, the reason I'm forced to save so frequently is that when I use Verl for GRPO, after about 30 steps or so, the model's training often enters a collapsed state at a certain step, with some metrics showing NaN (Not a Number). This seems similar to an issue others have mentioned, and even after trying their methods, it hasn't resolved the problem.

**Comment by syc1336** [comment]
how to convert the checkpoint to HF model files?

**Comment by Kirrito-k423** [comment]
中文：

这里的两个路径用途不同：

1. `trainer.resume_from_path` 用于继续同一次训练。它应该指向 verl checkpoint manager 保存的 step 目录，例如 `checkpoints/<project>/<experiment>/global_step_50`，而不是只包含 HF 权重的目录。当前代码也要求该路径存在且路径中包含 `global_step_`。
2. 对 FSDP 来说，建议保存/加载完整训练状态：`model`、`optimizer`、`extra`。如果只保存模型权重而没有 optimizer / lr scheduler / RNG / dataloader 等额外状态，就不能等价地“resume”完整训练；更接近于从某个权重重新开始训练。
3. `actor_rollout_ref.model.path` 是初始化模型权重用的路径，可以指向 base HF model 或一个已经 merge/export 出来的 HF 模型；它不会恢复 optimizer、scheduler、global step、RNG 等训练状态。
4. 如果只是想把 verl checkpoint 转成 HF 模型，可以用文档里的 merger：

```bash
python -m verl.model_merger merge \
  --backend fsdp \
  --local_dir checkpoints/<project>/<experiment>/global_step_50/actor \
  --target_dir /path/to/merged_hf_model
```

English:

These two paths serve different purposes:

1. `trainer.resume_from_path` is for continuing the same training run. It should point to the step directory saved by verl's checkpoint manager, for example `checkpoints/<project>/<experiment>/global_step_50`, not to a directory that only contains HF weights. The current code also requires the path to exist and contain `global_step_`.
2. For FSDP, the docs recommend saving/loading the full training state: `model`, `optimizer`, and `extra`. If only model weights are saved without optimizer / LR scheduler / RNG / dataloader-related state, it is not an equivalent full training resume; it is closer to starting a new run from those weights.
3. `actor_rollout_ref.model.path` is used to initialize model weights. It can point to a base HF model or a merged/exported HF model, but it does not restore optimizer, scheduler, global step, RNG, or other training state.
4. If the goal is only to convert a verl checkpoint to HF format, use the documented merger command:

```bash
python -m verl.model_merger merge \
  --backend fsdp \
  --local_dir checkpoints/<project>/<experiment>/global_step_50/actor \
  --target_dir /path/to/merged_hf_model
```

This response was generated by a scheduled Codex GPT5.5 issue processing task.
